### PR TITLE
Topic header

### PR DESCRIPTION
**Pull request by aronshamash**
Create a `topic-header` component within a new `topic` page component.

### With description
screen shot 2018-05-11 at 15 32 52

### No description
 screen shot 2018-05-11 at 15 33 01

## Chrome
topic-head

### CODE REVIEW

File: packages/topic/src/styles/shared.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+const sharedStyle = {
+  wrapper: {
+    width: "100%",
+    alignItems: "center",
+    flexDirection: "column",
+    paddingTop: spacing(6),
+    paddingBottom: spacing(8),
+    paddingLeft: spacing(2),
+    paddingRight: spacing(2),
```
Comment (jackharvey1) [inline packages/topic/src/styles/shared.js]: `paddingHorizontal: spacing(2)`?

File: packages/topic/src/topic-head.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import React, { Component } from "react";
+import { View } from "react-native";
+import { propTypes, defaultProps } from "./proptypes";
+import styles from "./styles";
+import HeadContent from "./topic-head-content";
+import Loading from "./topic-head-loading";
+
+const TopicHead = ({ name, description, isLoading }) => {
+  if (isLoading) {
+    return <Loading />;
```
Comment (jackharvey1) [inline packages/topic/src/topic-head.js]: Personally I find this style more readable: 

```
const TopicHead = ({ name, description, isLoading }) =>
  isLoading
    ? <Loading />
    : (<View style={styles.wrapper}>
        <HeadContent name={name} description={description} />
      </View>);
```
There's a few places you can do this in this PR. Up to you though.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/e30db9cbbd517610cb42a9385b71d790dc2a71a6/index.html

File: packages/topic/__tests__/android/__snapshots__/topic.test.js.snap [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      "paddingRight": 10,
+      "paddingTop": 30,
+      "width": "100%",
+    }
+  }
+>
+  <View
+    style={
+      Object {
+        "alignItems": "center",
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/topic/__tests__/android/__snapshots__/topic.test.js.snap]: React native already adds styles which set `display: flex`

File: packages/topic/src/styles/shared.js (hunk 1 above)
Comment (aronshamash) [reply packages/topic/src/styles/shared.js]: like it

File: packages/topic/src/topic-head.js (hunk 2 above)
Comment (aronshamash) [reply packages/topic/src/topic-head.js]: yes I agree

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/d959df694ca98724a8e82920061557535aaea9cf/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/bbb004868d7bb982a79c939d6df92b92295d24a2/index.html

File: packages/topic/__tests__/android/__snapshots__/topic.test.js.snap (hunk 3 above)
[the remark this answers is not in the document]
Comment (craigbilner) [reply packages/topic/__tests__/android/__snapshots__/topic.test.js.snap]: could we start as we mean to go on with snapshots and start making them smaller and more focused? For example you could test for the stylings in one snapshot and then mock them out to test for correct structure/rendering for example but as a generic `<Topic />` type look

[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/topic/__tests__/android/__snapshots__/topic.test.js.snap]: @craigbilner is there an example package that does this?

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/43136afd01ef9e503b262d835c236a7d2a7a55bb/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/9cbc59484f06017da2068340399edbddd2b84375/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/764ddf53e3e0c2981354a48a3127fa2d49983bc3/index.html

File: packages/topic/package.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  "scripts": {
+    "fmt": "prettier --write '**/*.*'",
+    "prettier:diff": "prettier --list-different '**/*.*'",
+    "lint": "eslint . && npm run prettier:diff",
+    "test:android": "jest --config='./__tests__/android/jest.config.js'",
+    "test:ios": "jest --config='./__tests__/ios/jest.config.js'",
+    "test:web": "jest --config='./__tests__/web/jest.config.js'",
+    "test:all": "yarn test:android --coverage & yarn test:ios --coverage & yarn test:web --coverage",
+    "dextrose-stories": "dextrose generate-stories $(pwd)",
+    "dextrose-clean": "dextrose clean-stories $(pwd)",
```
Comment (colincclark) [inline packages/topic/package.json]: these two scripts are now unrequired: `dextrose-stories` and `dextrose-clean`

File: packages/topic/src/name.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    testID="topic-name"
+    accessibilityLabel="topic-name"
+    accessibilityRole="heading"
+    style={styles.name}
+  >
+    {name}
+  </Text>
+);
+
+Name.propTypes = propTypes.name;
```
Comment (colincclark) [inline packages/topic/src/name.js]: this should be required

File: packages/topic/src/name.web.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  <ResponsiveName
+    testID="topic-name"
+    accessibilityLabel="topic-name"
+    accessibilityRole="heading"
+  >
+    {name}
+  </ResponsiveName>
+);
+
+Name.propTypes = propTypes.name;
```
Comment (colincclark) [inline packages/topic/src/name.web.js]: required

File: packages/topic/src/styles/index.ios.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { StyleSheet } from "react-native";
```
Comment (colincclark) [inline packages/topic/src/styles/index.ios.js]: these two native files can just be `index.js` and then web can be `index.web.js`....

File: packages/topic/src/styles/index.android.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { StyleSheet } from "react-native";
+import sharedStyles from "./shared";
+
+const webStyles = StyleSheet.create({
```
Comment (colincclark) [inline packages/topic/src/styles/index.android.js]: pls change variable name to `styles`

File: packages/topic/src/styles/index.ios.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+import sharedStyles from "./shared";
+
+const webStyles = StyleSheet.create({
```
Comment (colincclark) [inline packages/topic/src/styles/index.ios.js]: pls change variable name to styles

File: packages/topic/src/styles/index.js [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import { StyleSheet } from "react-native";
+import sharedStyles from "./shared";
+
+const webStyles = StyleSheet.create({
```
Comment (colincclark) [inline packages/topic/src/styles/index.js]: pls change variable name to styles

File: packages/topic/src/styles/responsive.web.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      padding-bottom: ${hasDescription ? spacing(7) : spacing(2)};
+    `,
+    wideUp: () => ` 
+      max-width: ${config.wideBpWidth};
+      padding-top: ${hasDescription ? spacing(3) : spacing(10)};
+      padding-bottom: ${hasDescription ? spacing(8) : spacing(3)};
+    `
+  });
+
+  HeadContainer.displayName = "HeadContainer";
```
Comment (colincclark) [inline packages/topic/src/styles/responsive.web.js]: the displayName can be passed as the last argument of the`withResponsiveStyles` method

File: packages/topic/src/proptypes.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+export const propTypes = {
+  name: PropTypes.string.isRequired,
+  description: PropTypes.string,
+  isLoading: PropTypes.bool
+};
+
+export const defaultProps = {
+  description: "",
+  isLoading: false
```
Comment (colincclark) [inline packages/topic/src/proptypes.js]: have a feeling isLoading should be `true` by default?

File: packages/topic/src/styles/responsive.web.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  fontSizes,
+  spacing
+} from "@times-components/styleguide";
+
+export const getHeadContainer = ({ hasDescription }) => {
+  const HeadContainer = withResponsiveStyles(View, {
+    base: () => `
+      width: 100%;
+      align-items: center;
+      flex-direction: column;
```
Comment (colincclark) [inline packages/topic/src/styles/responsive.web.js]: this is the default value in react-native so might be unrequired here?

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/c27fdeaefab3fe86c936353c65e9ca18cedea8ae/index.html

File: packages/topic/__tests__/android/__snapshots__/topic.test.js.snap (hunk 3 above)
[the remark this answers is not in the document]
Comment (colincclark) [reply packages/topic/__tests__/android/__snapshots__/topic.test.js.snap]: Slice and Article do snapshots fairly well...but I am not sure the value of changing these snapshots - they are pretty small and readable as is IMO

File: packages/topic/src/styles/shared.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    paddingBottom: spacing(4)
+  },
+  description: {
+    fontFamily: fonts.body,
+    textAlign: "center",
+    fontSize: fontSizes.tertiary,
+    lineHeight: 26,
+    color: colours.functional.primary,
+    paddingTop: spacing(4),
+    paddingBottom: spacing(4)
```
Comment (colincclark) [inline packages/topic/src/styles/shared.js]: `paddingVertical`? lol

File: packages/topic/src/name.js (hunk 5 above)
Comment (aronshamash) [reply packages/topic/src/name.js]: it is getting it from `proptypes.js` which is set to `PropTypes.string.isRequired`

File: packages/topic/src/name.web.js (hunk 6 above)
Comment (aronshamash) [reply packages/topic/src/name.web.js]: it is getting it from `proptypes.js` which is set to `PropTypes.string.isRequired`

File: packages/topic/src/styles/index.ios.js (hunk 7 above)
Comment (aronshamash) [reply packages/topic/src/styles/index.ios.js]: I am now thinking this is the most usual case and we should update the CLI?

File: packages/topic/src/styles/index.ios.js (hunk 9 above)
Comment (aronshamash) [reply packages/topic/src/styles/index.ios.js]: will raise CLI pr to sort this too.

File: packages/topic/src/styles/index.js (hunk 10 above)
Comment (aronshamash) [reply packages/topic/src/styles/index.js]: will raise CLI pr to sort this too.

File: packages/topic/src/name.web.js (hunk 6 above)
Comment (colincclark) [reply packages/topic/src/name.web.js]: a little confusing

File: packages/topic/src/proptypes.js (hunk 12 above)
Comment (aronshamash) [reply packages/topic/src/proptypes.js]: I was following pattern used in `article`

Comment (aronshamash) [reply packages/topic/src/proptypes.js]: which has `isLoading: false` as default

File: packages/topic/src/styles/responsive.web.js (hunk 11 above)
Comment (aronshamash) [reply packages/topic/src/styles/responsive.web.js]: nice

File: packages/topic/src/styles/responsive.web.js (hunk 13 above)
Comment (aronshamash) [reply packages/topic/src/styles/responsive.web.js]: it is unrequired

File: packages/topic/src/name.web.js (hunk 6 above)
Comment (aronshamash) [reply packages/topic/src/name.web.js]: ok can be if confused. I wanted to DRY it up as we already have it in `proptypes.js`

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/13c467163fc1a99c31ed886aff950d9ffffed32a/index.html

File: packages/topic/__tests__/android/__snapshots__/topic.test.js.snap (hunk 3 above)
[the remark this answers is not in the document]
Comment (craigbilner) [reply packages/topic/__tests__/android/__snapshots__/topic.test.js.snap]: they're very hard to read compared to a snapshot without the attribute noise which can be tested separately

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/fa8393b15d0ec01b7eb52236dc7da3e574017b8f/index.html

File: packages/topic/__tests__/react-native-mock-components.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
+import React from "react";
+
+const mockReactNativeComponents = (...componentNames) =>
```
Comment (colincclark) [inline packages/topic/__tests__/react-native-mock-components.js]: cool, but should be reusable by all packages

File: packages/topic/__tests__/shared-styling.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import "react-native";
+import "jest-styled-components";
+import React from "react";
+import renderer from "react-test-renderer";
+import Topic from "../src/topic";
+
+module.exports = () => {
```
Comment (colincclark) [inline packages/topic/__tests__/shared-styling.js]: `export default`

File: packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+      accessibilityRole="heading"
+      testID="topic-name"
+    >
+      Animals
+    </Text>
+  </View>
+</View>
+`;
+
+exports[`Topic functional tests on android should render correctly when loading 1`] = `<View />`;
```
Comment (colincclark) [inline packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap]: no value here :( would be good to have `TopicHeader` and `TopicHeaderLoading` components that can be mocked or just displayed here with a shallow render?

Comment (jimhunty) [reply packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap]: When these are not a separate package, how would you do that?

File: packages/topic/__tests__/react-native-mock-components.js (hunk 15 above)
Comment (jimhunty) [reply packages/topic/__tests__/react-native-mock-components.js]: Could go into utils package?

Comment (colincclark) [reply packages/topic/__tests__/react-native-mock-components.js]: yes, or jest configurator?

File: packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap (hunk 17 above)
Comment (aronshamash) [reply packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap]: this is done in `article` but from external packages, here they are not.

File: packages/topic/__tests__/react-native-mock-components.js (hunk 15 above)
Comment (aronshamash) [reply packages/topic/__tests__/react-native-mock-components.js]: `jest-configurator` as `utils` is not for development utils

File: packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap (hunk 17 above)
Comment (colincclark) [reply packages/topic/__tests__/android/__snapshots__/topic-functional.test.js.snap]: `jest.mock("../src/topic-head-loading", () => "TopicHeadLoading");`

File: packages/topic/__tests__/shared-functional.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
+import "react-native";
+import React from "react";
+import renderer from "react-test-renderer";
+import Topic from "../src/topic";
+
+jest.mock("../src/topic-head-loading", () => "TopicHeadLoading");
+jest.mock("../src/divider", () => "Divider");
+
+module.exports = () => {
```
Comment (colincclark) [inline packages/topic/__tests__/shared-functional.js]: `export default`

**Pull request merged**